src/Compilers/Core/Portable/InternalUtilities/SemaphoreExtensions.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Roslyn.Utilities
+{
+    internal static class SemaphoreExtensions
+    {
+        public static SemaphoreDisposer DisposableWait(this Semaphore semaphore, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (cancellationToken.CanBeCanceled)
+            {
+                int signalledIndex = WaitHandle.WaitAny(new[] { semaphore, cancellationToken.WaitHandle });
+                if (signalledIndex != 0)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    throw ExceptionUtilities.Unreachable;
+                }
+            }
+            else
+            {
+                semaphore.WaitOne();
+            }
+
+            return new SemaphoreDisposer(semaphore);
+        }
+
+        public static Task<SemaphoreDisposer> DisposableWaitAsync(this Semaphore semaphore, CancellationToken cancellationToken = default(CancellationToken))
```

Comment:
@jasonmalinowski @bradwilson I addressed the thread pool concern by using dedicated threads, and the resource limit issue by verifying the default behavior of xUnit with respect to parallelism. The current implementation is not ideal, but it's reliable. I don't believe it will be a problem in the future either due to the availability and more frequent use of `SemaphoreSlim`.